L3 UDN: EgressIP hosted by primary interface (`breth0`)

martinkennelly opened the pull request:
Please read the enhancement first: https://github.com/ovn-org/ovn-kubernetes/pull/4438

- [X] Add e2es 
- [X] Add unit tests for EIP OVN controller 
- [X] `syncPodAssignmentCache` refactor for L3 UDN
- [X] `syncStaleSNATRules` refactor for L3 UDN
- [x] Documentation update
- [x] Convert the logs to be network aware

pperiyasamy commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -1644,7 +1674,7 @@ func (e *egressIPZoneController) addPodEgressIPAssignment(egressIPName string, s
 	// don't add a reroute policy if the egress node towards which we are adding this doesn't exist
 	if loadedEgressNode && loadedPodNode && isLocalZonePod {
-		ops, err = e.createReroutePolicyOps(ops, podIPs, status, egressIPName, nextHopIP)
+		ops, err = e.createReroutePolicyOps(ops, podIPs, status, mark, egressIPName, nextHopIP)
```
isn't this redunant to do pkt marking here ? this is already done in GW router, also mark doesn't survive when it goes out to wire.

martinkennelly replied:
I was thinking of EIP multi NIC when I wrote this where it can reroute to local mgnt port.

pperiyasamy reviewed (commented):
reviewed until 93e13cbc47065335ec9aec6d39cf9b73f3d36a6b.

pperiyasamy commented on go-controller/pkg/node/gateway_egressip.go:
```diff
@@ -0,0 +1,542 @@
+}
+
+func (g *bridgeEIPAddrManager) syncEgressIP(objs []interface{}) error {
+	// caller must synchronise
```
typo: synchronize

martinkennelly replied:
its the european way of spelling it :)

pperiyasamy commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -1440,6 +1441,21 @@ func commonFlows(subnets []*net.IPNet, bridge *bridgeConfiguration) ([]string, e
+				if util.IsNetworkSegmentationSupportEnabled() && config.OVNKubernetesFeature.EnableInterconnect &&
+					config.Gateway.Mode != config.GatewayModeDisabled && bridge.eipMarkIPs != nil {
+					if netConfig.masqCTMark != ctMarkOVN {
+						for mark, eip := range bridge.eipMarkIPs.GetIPv4() {
```
This is really nice! having eipMarkIPs as source of truth to program SNAT OF rules.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -188,7 +188,18 @@ func (bnc *BaseNetworkController) reconcileEgressIP(old, new *egressipv1.EgressI
 			}
+			var role string
 			for _, namespace := range namespaces {
+				// determine if this network configures EIP for this namespace
```
I thought about that but i didnt want a wrapper around this as i thought this was easier to read. Id still need to check for error, return err if needed and call continue if not primary  and no error. WDYT? I think id rather leave it .

martinkennelly replied:
I think its ok and I need to lookup the active network, so ill mark this as resolved.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -2309,3 +2381,14 @@ func getPodNamespaceAndNameFromKey(podKey string) (string, string) {
+	if n2 != nil {
+		return n2
+	}
+	panic("getNonNilNamespace(): received two nil Namespace pointers")
```
we should never get into this scenario where theres two nil namespaces passed and if we do I wanted to scream via a panic instead of it possibly hidden in error logs. WDYT?

martinkennelly replied:
Agree, crashing the controller is harsh. Removed - not needed anymore

martinkennelly commented on go-controller/pkg/util/egressip_annotation.go:
```diff
@@ -0,0 +1,69 @@
+
+// EgressIPMarkAnnotationChanged returns true if the EgressIP mark annotation changed
+func EgressIPMarkAnnotationChanged(annotationA, AnnotationB map[string]string) bool {
+	return annotationA[EgressIPMarkAnnotation] != AnnotationB[EgressIPMarkAnnotation]
```
it is handling if its not set. This func just informs if the annotation has changed from either not set to set or vice versa.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -613,14 +613,26 @@ func (bnc *BaseNetworkController) addPodEgressIPAssignments(name string, statusA
+		if err != nil {
+			return fmt.Errorf("failed to get active network for Namespace %s: %v", pod.Name, err)
+		}
+		nadNames := ni.GetNADs()
```
There can be multiple nads referencing the same network.

martinkennelly commented on go-controller/pkg/node/gateway_egressip.go:
```diff
@@ -0,0 +1,542 @@
+		}
+		// must always add to cache before adding IP because we want to inform node ip handler that this is not a valid node IP
+		g.cache.insertMarkIP(pktMark, ip)
+		if err = g.addIPToAnnotation(ip); err != nil {
```
If the EIP is deleted while controller is down, we've no way to know what addresses we configured on the bridge and clean it up.

martinkennelly replied:
Maybe theres a better idea that I am not thinking of to get the set of IPs we added but I couldnt think of it.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -382,8 +388,12 @@ func (bnc *BaseNetworkController) reconcileEgressIPNamespace(old, new *v1.Namesp
+		}
 		if !namespaceSelector.Matches(oldLabels) && namespaceSelector.Matches(newLabels) {
-			if err := bnc.addNamespaceEgressIPAssignments(egressIP.Name, egressIP.Status.Items, newNamespace, egressIP.Spec.PodSelector); err != nil {
+			if err := bnc.addNamespaceEgressIPAssignments(egressIP.Name, egressIP.Status.Items, mark, newNamespace, egressIP.Spec.PodSelector); err != nil {
```
I was trying to only pass the info that was required and tried not to pass the whole map. I thought that was more readable for folks reading the func signatures and deduce what fn argument was used for what.

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -1723,3 +1711,149 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+			if err != nil {
+				return fmt.Errorf("failed to get mark from EgressIP %s: %v", egressIP.Name, err)
+			}
+			if !mark.IsValid() {
```
It should be required as on startup we reserve any IDs that are already allocated and then we can depend on the ID allocator giving us non-used IDs.

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -1723,3 +1711,149 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+	if mark < 0 {
+		return fmt.Errorf("unable to reserve mark because calculated offset is less than zero")
+	}
+	if err := eIPC.markAllocator.ReserveID(name, mark); err != nil {
```
no need - ID allocator takes care of it - we pass a max

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -4049,11 +4208,32 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 		})
 	})
+
+	ginkgo.Context("EgressIP Mark cache", func() {
```
I wrote one and tested it but it took a long time to iterate through all the range values - something like 2 minutes on my local PC. I actually dont want it to return an error if range is exhausted because i dont want to block EIP assignment for pods assigned to CDN. I tested to make sure it doesnt return an error for that scenario.

maiqueb replied:
Can't you have a way to inject a max value to the range for the tests ? (really asking)

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -3074,6 +3084,121 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
+					if err != nil {
+						return err
+					}
+					eIP.Annotations = map[string]string{}
```
will do - it must be original mark - marks cannot change for the life time of the EIP object.

martinkennelly commented:
@pperiyasamy I added some new commits but made no extra changes to the commits you viewed except addressing your comments and it wasnt a large amount of code. yet. Any comments Ive marked resolved, I changed the code. If you disagree, mark them unresolved.

martinkennelly commented:
Rebase taking longer than expected to incorporate patryks changes to enable "udn enabled services" from the CDN network. Should be done today if all goes well.

maiqueb commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -1723,3 +1711,149 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+// to build the initial mark cache
+func (eIPC *egressIPClusterController) getOrAllocMark(name string) (int, error) {
+	if name == "" {
+		panic("EIP getOrAllocMark(): name cannot be blank")
```
+1 ; if the method's signature has an error, well, let's return it :)

maiqueb commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -396,6 +365,8 @@ func newEgressIPController(ovnClient *util.OVNClusterManagerClientset, wf *facto
 		EIPClient:          ovnClient.EgressIPClient,
 		CloudNetworkClient: ovnClient.CloudNetworkClient,
 	}
+	markAllocator, _ := getEgressIPMarkAllocator()
```
why do you discard the error ? shouldn't you handle it ?

martinkennelly replied:
ya, true... i should refactor the func instead of covering it up..

martinkennelly replied:
Done

maiqueb reviewed (commented):
reviewed up to (excluding) "EIP CM: add pkt mark to annotations".

martinkennelly commented on go-controller/pkg/libovsdb/ops/router.go:
```diff
@@ -256,6 +256,27 @@ func FindLogicalRouterPoliciesWithPredicate(nbClient libovsdbclient.Client, p lo
 
+// FindALogicalRouterPoliciesWithPredicate looks up a logical router policies from
+// the cache based on a given predicate
+func FindALogicalRouterPoliciesWithPredicate(nbClient libovsdbclient.Client, routerName string, p logicalRouterPolicyPredicate) ([]*nbdb.LogicalRouterPolicy, error) {
```
it should reduce the number of CPU cycles in large clusters because we only focus on LRPs attached to a specific router.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -62,24 +65,64 @@ func getEgressIPAddrSetDbIDs(name egressIPAddrSetName, controller string) *libov
 
-func getEgressIPLRPNoReRouteDbIDs(priority int, uniqueName egressIPNoReroutePolicyName, ipFamily egressIPFamilyValue) *libovsdbops.DbObjectIDs {
-	return libovsdbops.NewDbObjectIDs(libovsdbops.LogicalRouterPolicyEgressIP, DefaultNetworkControllerName, map[libovsdbops.ExternalIDKey]string{
+func getEgressIPLRPReRouteDbIDs(egressIPName, podNamespace, podName string, ipFamily egressIPFamilyValue, controller string) *libovsdbops.DbObjectIDs {
```
yep, I did it for NATs (because i didnt want to have to know the pod IPs) and all LRPs.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -1108,10 +1153,15 @@ func (bnc *BaseNetworkController) syncPodAssignmentCache(egressIPCache map[strin
+		if item.Priority != types.EgressIPReroutePriority || item.ExternalIDs[libovsdbops.OwnerControllerKey.String()] != bnc.controllerName {
+			return false
+		}
+		objReferences := strings.Split(item.ExternalIDs[libovsdbops.ObjectNameKey.String()], "_")
```
Removed this ugly code.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip.go:
```diff
@@ -647,6 +647,9 @@ func (bnc *BaseNetworkController) addPodEgressIPAssignments(name string, statusA
 		klog.Infof("Pod %s is already in completed state, skipping egress ip assignment", podKey)
 		return nil
 	}
+	if util.PodWantsHostNetwork(pod) {
```
remove it not required.

martinkennelly commented on go-controller/pkg/ovn/base_network_controller_egressip_udn_l3_test.go:
```diff
@@ -0,0 +1,2732 @@
+					egressSVCServedPodsASUDNv4,
+					egressIPServedPodsASUDNv4,
+					egressNodeIPsASUDNv4,
+				}
```
this is OVN unit tests - we dont program UDN bridge NAT flows here.

martinkennelly commented on go-controller/pkg/ovn/secondary_layer3_network_controller.go:
```diff
@@ -387,6 +508,14 @@ func NewSecondaryLayer3NetworkController(cnci *CommonNetworkControllerInfo, netI
 func (oc *SecondaryLayer3NetworkController) initRetryFramework() {
 	oc.retryPods = oc.newRetryFramework(factory.PodType)
 	oc.retryNodes = oc.newRetryFramework(factory.NodeType)
+	if config.OVNKubernetesFeature.EnableEgressIP && config.OVNKubernetesFeature.EnableInterconnect &&
```
Removed the condition.

tssurya commented:
Is there a way to split this into smaller PRs or any tips for reviewing or shall I just to 1 to 17 here.. :D

martinkennelly commented:
> Is there a way to split this into smaller PRs or any tips/tricks for reviewing that I need to know or shall I just to 1 to 17 here.. :D

1 to 17 unfortunately.

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1875,9 +1967,10 @@ func (e *egressIPZoneController) deleteReroutePolicyOps(ops []ovsdb.Operation, p
+func (e *EgressIPController) deleteEgressIPStatusSetup(name string, status egressipv1.EgressIPStatusItem) ([]net.IP, error) {
 	var err error
 	var ops []ovsdb.Operation
+	defaultNetwork := util.DefaultNetInfo{}
```
I see this is repeating everywhere.... I think its perhaps better to store a netInfo object in the EIP controller? or is this changing in future commits..?

martinkennelly replied:
I dont think it would increase readability and it seems silly to just add that struct to the eip controller struct when i can just instantiate it. Its basically used as a constant.... so i guess i could make it a constant? It would make things slightly less readable but concise. Up to you..

tssurya commented on go-controller/pkg/ovn/base_network_controller.go:
```diff
@@ -861,6 +871,58 @@ func (bnc *BaseNetworkController) GetNetworkRole(pod *kapi.Pod) (string, error)
+// NOTE: Like in other places, expectation is this function is always called
+// from controller's that have some relation to the given namespace, unrelated
+// networks are treated as secondary networks so caller has to be careful
+func (bnc *BaseNetworkController) GetNetworkRoleForNamespace(ns *kapi.Namespace) (string, error) {
```
I think what you want is "GetActiveNetworkForNamespace" ? @trozet did several scale improvements along this... so maybe moving to getactivenetworkfornamespace is the right thing here? so maybe you can drop this commit - but let me see how you use this...

martinkennelly replied:
This commit was supposed to be dropped by me - oops. Its not even used in the current implementation. Removing full commit.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -151,15 +151,15 @@ func (eIPC *egressIPClusterController) patchReplaceEgressIPStatus(name string, s
 }
 
-type allocator struct {
+type nodeAllocator struct {
```
also add a doc string for that this allocator does? :) I know we never had it before but now that you are cleaning it up.....

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
can you add more to the commit message? I forget what we decided in how we wanted to communicate the mark across components...

martinkennelly replied:
Added:
```
This packet mark will be consumed by OVN Kubernetes
controllers running on each Kubernetes nodes.

EgressIP OVN controller	will mark packets from pods
attached to a role primary UDN and selected by an EgressIP
will get packet	marked by OVN LRPs on the egress nodes gateway
router.

EgressIP node gateway controller will add flows	to the external
bridge to SNAT packets that match the packet marks as defined
in the EgressIP	CR to a	specific EgressIP IP.

For an egress IP assigned to a host secondary interface, the 
packet mark will also be used in LRPs to redirect packets
to an egress nodes managements port and	IP rules and IPTables
rules will match on this packet	mark and SNAT to an egress IP.

Debugging incorrect ovs	flows or OVN constructs	will become easier
as we can inspect the EgressIP CR to gather the	expected packet	mark
for a given EgressIP.

```

martinkennelly replied:
LMK if its not sufficient.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -1723,3 +1711,146 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+// log an error instead of failing because we do not wish to block primary default network egress IP assignments due to potential
+// mark range exhaustion. Primary default network egress IP currently does not utilize marks to config EgressIP.
+// Generating the status patch is mandatory
+func (eIPC *egressIPClusterController) generateEgressIPPatches(name string, annotations map[string]string,
```
can you expand on why the mark logic is coupled with status changes?
I am confused as to why this is being called from every status update part.... I'd expect those bits to not change...

martinkennelly replied:
> can you expand on why the mark logic is coupled with status changes?

its to ensure when theres any status changes theres always going to be a mark, its a no-op usually. I was too lazy to do a lookup first to see if we need to perform the op.

martinkennelly replied:
> I don't understand this; do you mean the cluster default network ? (i.e. the kubernetes managed network)

yes

> And what do we do if/when we the exhaust the mark range ? IMHO if the workload has required egressIPs, we need to put it in a crash-loop if we cannot fulfill the user's intents.

it was a design choice - CDN pods selected by an EIP will continue to function as it doesnt depend on pkt marks but UDN pods will not. I didnt want to block CDN pods.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -1723,3 +1711,146 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+}
+
+func createAnnotWithMark(mark int) map[string]string {
+	return map[string]string{util.EgressIPMarkAnnotation: fmt.Sprintf("%d", mark)}
```
Is this a mark per EIP? or mark per EIP Object?  - can you add the design logic to some place if its already covered in @pperiyasamy 's design PR then its fine, just want to make sure its detailed somewhere
maybe in the commit message?: when do you set the mark, update, delete it etc

martinkennelly replied:
Ive updated the docs to show how the mark is used and its also in the design doc. I ensured thats updated yesterday.

martinkennelly replied:
>Is this a mark per EIP? or mark per EIP Object?

Per EIP object. The reason is only one EIP IP can be assigned per node. No need to add a mark per EIP IP.

martinkennelly replied:
Unresolve this if youve more Qs.

tssurya commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -1723,3 +1711,146 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+}
+
+// syncEgressIPMarkAllocator iterates over all existing EgressIPs. It builds a mark cache of existing marks stored on each
+// EgressIP annotation or allocates and adds a new mark to an EgressIP if it doesn't exist
```
when you allocate and add a new mark how do you base the algorithm off the base?

martinkennelly replied:
I dont think I understand your question but..
Heres how it works.
CM starts up and reads all the EIP objects from the kapi and builds a cache of already allocated marks. It then can assign new marks that do not conflict with old marks.

tssurya commented on go-controller/pkg/util/egressip_annotation.go:
```diff
@@ -0,0 +1,69 @@
+func (em EgressIPMark) IsAvailable() bool {
+	return em.strValue != ""
+}
+
```
why is the add mark annotation not part of util and part of the patch logic in EIP controller?

martinkennelly replied:
its shared logic so i stuffed it in util - I hate there is a util in ovn-k.. its an anti pattern.
EIP OVN controller handles the marks so i thought id make that uniform between CM and OVN EIP controller. I think EIP breth0 controller also needs to handle marks. I think wrapping the mark logic around this `EgressIPMark` struct makes it easier to ready the code.

tssurya replied:
ack

tssurya reviewed (commented):
I got through the first 4 commits... I have some high level design doubts first which I need to clear with you first to continue the review... - maybe if I went though all 17 I might understand it better or if I huddled with you...
1) mark logic
2) controller logic
the commit splits make it easy to review so no complaints there thank you for doing that!

martinkennelly commented:
Updated enhancement which may help with some higher level questions; https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4438

martinkennelly commented on go-controller/pkg/ovn/egressip.go:
ya, its not fully seperated because id have to setup the handler independently. Its ready for that if we ever want to make the CDN optional.

martinkennelly replied:
cluster default network

martinkennelly commented on go-controller/pkg/network-controller-manager/network_controller_manager.go:
```diff
@@ -48,13 +49,15 @@ type NetworkControllerManager struct {
-
+	stopChan                 chan struct{}
+	wg                       *sync.WaitGroup
+	portCache                *ovn.PortCache
```
its a shared cache now where as before it was not. I needed to do this so that EIP controller can lookup ports from all networks. I was a little concerned with this because of lock contention but i think it should be ok.

The shared port cache is at this level because NCM controls the creation of new networks and needs to know where the cache is located in-order to deliver this info to the networks it creates and therefore eip controller has access to all the networks ports.

martinkennelly replied:
@npinaeva can you ack youve seen this?

npinaeva replied:
yep, that should be fairly simple to make it lockable per-pod-name, @martinkennelly ping me if you need help/reviews

martinkennelly replied:
Note: create an issue to track this and self assign.

martinkennelly commented on go-controller/pkg/network-controller-manager/network_controller_manager.go:
```diff
@@ -48,13 +49,15 @@ type NetworkControllerManager struct {
 	// net-attach-def controller handle net-attach-def and create/delete network controllers
 	nadController *nad.NetAttachDefinitionController
+	// eIPController programs OVN to support EgressIP
+	eIPController *ovn.EgressIPController
```
>So EIP is going to be global for all networks right? Meaning the given EIP object created by the admin will span all networks across the cluster? 

yes.

>are planning to have a EIP controller per network and ensure network-1's controller programmes the EIP OVN objects for namespace-1 and vice versa for network-2?

No, one EIP controller programs all networks.

tssurya replied:
ack can you clarify this in code comments saying 1 global controller does the programming for all the networks? cause different features did the design differently so I am always confused with this point...

martinkennelly replied:
sure - thanks for highlighting this! Ill add it.

martinkennelly commented on go-controller/pkg/network-controller-manager/network_controller_manager.go:
```diff
@@ -406,6 +409,10 @@ func (cm *NetworkControllerManager) Start(ctx context.Context) error {
 		}
 	}
+	if config.OVNKubernetesFeature.EnableEgressIP {
+		cm.eIPController = ovn.NewEIPController(cm.nbClient, cm.kube, cm.watchFactory, cm.recorder, cm.portCache, cm.nadController,
```
it needs nad controller to lookup networks efficiently.

martinkennelly commented:
Forced pushed:
* Rebased to main
* Removed a commit that added code that wasnt even used -`GetNetworkRoleForNamespace()`
* Added fn doc text after addressing @tssurya comments

maiqueb commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -4047,11 +4206,32 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
+	ginkgo.Context("EgressIP Mark cache", func() {
+		ginkgo.It("should round robin when mark range is exhausted", func() {
+			iterations := (util.EgressIPMarkMax - util.EgressIPMarkBase) * 2 // iterate over the allowed range
+			nodeAlloc, _ := getEgressIPMarkAllocator()
```
shouldn't we assert an error is not thrown here ?

martinkennelly replied:
done

maiqueb reviewed (commented):
Reviewed "EIP CM: add pkt mark to annotations".

Will keep reviewing ... (sorry for the long hiatus)

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -974,6 +948,20 @@ func (eIPC *egressIPClusterController) reconcileEgressIP(old, new *egressipv1.Eg
 		statusToRemove = append(statusToRemove, status)
 		ipsToRemove.Insert(status.EgressIP)
 	}
+	// Adding the mark to annotations is bundled with status update in-order to minimise updates, cover the case where there is no update to status
+	// and mark annotation has been modified / removed. This should only occur for an update and the mark was previous set.
+	if old != nil && new != nil {
+		if ipsToAssign.Len() == 0 && ipsToRemove.Len() == 0 && util.IsEgressIPMarkSet(old.Annotations) &&
+			util.EgressIPMarkAnnotationChanged(old.Annotations, new.Annotations) {
```
done

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -3072,6 +3082,121 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
+
+		ginkgo.It("should manage mark cache for add/update/delete", func() {
+			app.Action = func(ctx *cli.Context) error {
+				egressIP := "192.168.126.10"
+				node1IPv4 := "192.168.126.12/24"
+				node2IPv4 := "192.168.126.51/24"
```
not worth the effort imo

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -3072,6 +3082,121 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
+				gomega.Expect(err).ShouldNot(gomega.HaveOccurred(), "Deletion of EIP must succeed")
+				// if reserve ID succeeds, then the previous entry was removed.
+				gomega.Eventually(func() error {
+					return fakeClusterManagerOVN.eIPC.markAllocator.ReserveID(eIP.Name, 9999)
```
Ill add the test.

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -3072,6 +3082,121 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
+				}).Should(gomega.Succeed())
+				return nil
+			}
+			err := app.Run([]string{app.Name})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
```
not going to bother changing because its more readable as it is right now imo.

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -3855,6 +3980,38 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
+				gomega.Expect(eIPMark).Should(gomega.Equal(eIPMarkFromCache))
+				eIP2MarkFromCache, err := fakeClusterManagerOVN.eIPC.getOrAllocMark(egressIPName2)
+				gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+				gomega.Expect(eIP2Mark).Should(gomega.Equal(eIP2MarkFromCache))
```
done

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller_test.go:
```diff
@@ -3072,6 +3082,121 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
+				node1IPv4 := "192.168.126.12/24"
+				node2IPv4 := "192.168.126.51/24"
+
+				node1 := v1.Node{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: node1Name,
+						Annotations: map[string]string{
+							"k8s.ovn.org/node-primary-ifaddr": fmt.Sprintf("{\"ipv4\": \"%s\", \"ipv6\": \"%s\"}", node1IPv4, ""),
+							"k8s.ovn.org/node-subnets":        fmt.Sprintf("{\"default\":[\"%s\", \"%s\"]}", v4NodeSubnet, v6NodeSubnet),
+							util.OVNNodeHostCIDRs:             fmt.Sprintf("[\"%s\"]", node1IPv4),
+						},
+						Labels: map[string]string{
+							"k8s.ovn.org/egress-assignable": "",
+						},
+					},
+					Status: v1.NodeStatus{
+						Conditions: []v1.NodeCondition{
+							{
+								Type:   v1.NodeReady,
+								Status: v1.ConditionTrue,
+							},
+						},
+					},
+				}
```
i am following the existing style for egress ip unit tests.

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -742,18 +853,22 @@ func (e *EgressIPController) deleteEgressIPAssignments(name string, statusesToRe
 				continue
 			}
+			podNamespace, podName := getPodNamespaceAndNameFromKey(podKey)
+			ni, err := e.nadController.GetActiveNetworkForNamespace(podNamespace)
```
this may not work if the namespace is gone. You may want to store the active network in the pod assignment cache at add time.

martinkennelly replied:
is the network is gone, then all the OVN constructs to support EgressIP are removed. Ill add a unit test cases for this.

martinkennelly replied:
I added a commit on top to address this.

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -941,12 +1049,23 @@ type egressIPCacheEntry struct {
-	egressLocalNodes sets.Set[string]
+}
+
+type redirectIPs struct {
```
some comments here please

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1005,106 +1127,321 @@ func (e *EgressIPController) syncLocalNodeZonesCache() error {
 // all the pods that are managed by egressIPs.
 // NOTE: This is done mostly to handle the corner case where one pod has more than one
 // egressIP object matching it, in which case we do the ovn setup only for one of the objects.
 // This corner case  of same pod matching more than one object will not work for IC deployments
```
while you are here, should this say:

This corner case of the same pod matching more than one object will is not applicable in IC deployments

?

martinkennelly replied:
it already says this - i dont understand

trozet replied:
I rephrased it and got rid of a space. The rephrase was just to say the case is not relevant to IC deployments rather than "something is broken/not work" in IC deployments. not a big deal though if you dont fix that in your PR.

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1005,106 +1127,321 @@ func (e *EgressIPController) syncLocalNodeZonesCache() error {
-			for _, podIP := range podIPsSet.UnsortedList() {
-				podIPs = append(podIPs, net.ParseIP(podIP))
+func (e *EgressIPController) syncPodAssignmentCache(egressIPCache egressIPCache) error {
+	//routerName := bnc.GetNetworkScopedClusterRouterName()
```
remove

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1005,106 +1127,321 @@ func (e *EgressIPController) syncLocalNodeZonesCache() error {
 // since internal cache based logic will be different for different ovnkube-controllers
 // zone can think objA is active while zoneb can think objB is active if both have multiple choice options
-func (e *EgressIPController) syncPodAssignmentCache(egressIPCache map[string]egressIPCacheEntry) error {
-	e.podAssignmentMutex.Lock()
```
why do you no longer need to hold the mutex here?

martinkennelly replied:
it doesnt need to be held because this fn is executed before the handlers are run but i shouldnt have removed it in that commit - should be done in a separate commit. Put it back.

trozet reviewed (changes requested):
made it up to generateCacheForEgressIP in  [EIP OVN: make controller multi net aware & consider pkt mark](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4530/commits/4dd0d757a0dff5da04745c0fbe2e41b7e51eda40)

will continue tmrw

martinkennelly commented on go-controller/pkg/ovn/controller/udnenabledsvc/udn_enabled_svc.go:
```diff
@@ -0,0 +1,299 @@
+
+// NewController creates a new controller to sync UDN enabled services clusterIPs with an OVN address set. Address set is
+// never deleted and maybe read-only referenced by consumers. Single worker only for processing events.
+func NewController(nbClient libovsdbclient.Client, addressSetFactory addressset.AddressSetFactory,
```
we need to skip processing of the routers LRPs for egress IP if the dst ip is a udn enabled svc IP

martinkennelly commented on go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go:
```diff
@@ -231,8 +231,8 @@ func (c *Controller) Run(wg *sync.WaitGroup, threadiness int) error {
 	}
-
-	err = c.initClusterEgressPolicies(c.nbClient, c.addressSetFactory, c.controllerName, c.GetNetworkScopedClusterRouterName())
+	subnets := util.GetAllClusterSubnetsFromEntries(c.Subnets())
```
ya, egress service consumes funcs belonging to EIP - I hate it.

martinkennelly commented on go-controller/pkg/util/egressip_annotation.go:
```diff
@@ -0,0 +1,69 @@
+const (
+	EgressIPMarkAnnotation = "k8s.ovn.org/egressip-mark"
+	EgressIPMarkBase       = 50000
+	EgressIPMarkMax        = math.MaxUint16
```
There is no limit currently on the number of EIPs. Ill try to find a reasonable limit.

martinkennelly replied:
LMK if youre unhappy with that range.

martinkennelly commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -770,7 +770,7 @@ func (e *EgressIPController) deleteEgressIPAssignments(name string, statusesToRe
 				// NOTE(tssurya): There is no way to infer if pod was local to this zone or not,
 				// so we try to nuke the IP from address-set anyways - it will be a no-op for remote pods
-				if err := e.deletePodIPsFromAddressSet(podIPs); err != nil {
+				if err := e.deletePodIPsFromAddressSet(podStatus.podIPs); err != nil {
```
it comes later

martinkennelly commented on go-controller/pkg/libovsdb/ops/router.go:
```diff
@@ -257,6 +257,27 @@ func FindLogicalRouterPoliciesWithPredicate(nbClient libovsdbclient.Client, p lo
 
+// FindALogicalRouterPoliciesWithPredicate looks up a logical router policies from
+// the cache based on a given predicate
+func FindALogicalRouterPoliciesWithPredicate(nbClient libovsdbclient.Client, routerName string, p logicalRouterPolicyPredicate) ([]*nbdb.LogicalRouterPolicy, error) {
```
Its already taken i believe - i think the names ok - i am trying to signal im searching for LRPs attached to a single router.

martinkennelly commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -806,22 +871,67 @@ func (oc *DefaultNetworkController) deletePodEgressIPAssignments(name string, st
+			isLocalZonePod = false
+		}
+	}
+	klog.Infof("## martin isPodScheduledinLocalZone() pod %s with node %s is local? %v", pod.Name, pod.Spec.NodeName, isLocalZonePod)
```
oops - cleaned

martinkennelly commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -195,7 +195,6 @@ func newDefaultNetworkControllerCommon(cnci *CommonNetworkControllerInfo,
 	if err != nil {
 		return nil, fmt.Errorf("unable to create new admin policy based external route controller while creating new default network controller :%w", err)
 	}
-	eIPController.zone = cnci.zone
```
mistake - cleaned.

trozet commented on test/e2e/egressip.go:
```diff
@@ -2705,6 +2705,124 @@ spec:
 			"and verify the expected IP, failed for EgressIP %s: %v", egressIPName, err)
 	})
+
+	ginkgo.DescribeTable("[OVN network] multiple namespaces with different primary networks", func(otherNetworkAttachParms networkAttachmentConfigParams) {
```
nice

trozet commented on go-controller/pkg/kubevirt/external_ids.go:
```diff
@@ -33,6 +33,11 @@ func externalIDsContainsVM(externalIDs map[string]string, vm *ktypes.NamespacedN
 	if vm == nil {
 		return false
 	}
+	// FIXME: VM IDs have no DB IDs and therefore may clash with other LRPs that do contain DB IBs. They will always have ObjectNameKey
```
@qinqon fyi

trozet reviewed (changes requested):
finished reviewing, nice work @martinkennelly 

I didn't review all of the unit tests, and I also didn't review 1 commit which I asked @npinaeva to take a look at.

martinkennelly commented on go-controller/pkg/clustermanager/egressip_controller.go:
```diff
@@ -1726,3 +1706,167 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+				return fmt.Errorf("failed to reserve mark for EgressIP %s: %v", egressIP.Name, err)
+			}
+		} else {
+			mark, err := eIPC.getOrAllocMark(egressIP.Name)
```
duh - dumb - thanks man!

martinkennelly commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -403,12 +511,21 @@ func (e *EgressIPController) reconcileEgressIPNamespace(old, new *v1.Namespace)
 		}
 		if namespaceSelector.Matches(oldLabels) && !namespaceSelector.Matches(newLabels) {
-			if err := e.deleteNamespaceEgressIPAssignment(egressIP.Name, egressIP.Status.Items, oldNamespace, egressIP.Spec.PodSelector); err != nil {
+			ni, err := e.nadController.GetActiveNetworkForNamespace(namespaceName)
```
True - i added a commit on top to address this.

martinkennelly commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1268,43 +1652,137 @@ func (e *EgressIPController) syncStaleSNATRules(egressIPCache map[string]egressI
-func (e *EgressIPController) generateCacheForEgressIP() (map[string]egressIPCacheEntry, error) {
-	egressIPCache := make(map[string]egressIPCacheEntry)
-	egressIPs, err := e.watchFactory.GetEgressIPs()
+func (e *EgressIPController) generateCacheForEgressIP() (egressIPCache, error) {
```
If the UDN is deleted we dont have to worry as the config will be removed.
If the UDN is present, I think its ok because the secondary controllers dont depend on the EIP controller to be started. 

> or does this work because egress IP controller is not started until the retry framework starts in DNC? If so, do we end up missing any events or scenarios when secondary network controllers have already come up and started.

I cannot think of any.

martinkennelly replied:
its also possible i fully extract the EIP controller from DNC and avoid all this. The code is setup for that now, just a little more effort tim.

maiqueb replied:
IMHO it would be preferable to extract the EIP controller for the default network ... 

But I don't understand how much more effort it would take, and if that can be squeezed into your deadline ...

trozet replied:
"If the UDN is present, I think its ok because the secondary controllers dont depend on the EIP controller to be started."

So I see the secondary l3 controller calls: ensureL3ClusterRouterPoliciesForNetwork, which relies on nodeZoneState cache, which is updated during syncEgressIPs:

```
	// update localZones cache of eIPCZoneController
	// WatchNodes() is called before WatchEgressIPNamespaces() so the oc.localZones cache
	// will be updated whereas WatchEgressNodes() is called after WatchEgressIPNamespaces()
	// and so we must update the cache to ensure we are not stale.
	if err := e.syncLocalNodeZonesCache(); err != nil {
		return fmt.Errorf("syncLocalNodeZonesCache unable to update the local zones node cache: %v", err)
	}
```

I think we need to fix this. If its easy for you to rip out the EIP controller from DNC and start it separately that is good. Otherwise we can explore just abstracting out the call to the sync function to run before the controllers are started, and remove it from syncing in the DNC retry framework. That should be very easy to do.

martinkennelly replied:
I apologise for that comment to you. I should have thought about that interaction between the two components.
I explored both paths and I would like to extract the EIP controller fully from the DNC in a follow up PR.
I added a commit on top to sync the OVN zone cache.

martinkennelly commented on go-controller/pkg/node/gateway_egressip.go:
```diff
@@ -0,0 +1,521 @@
+	})
+}
+
+func (g *bridgeEIPAddrManager) addIPBridge(ip net.IP) error {
```
> why are we adding egress IP to the bridge here? Is it so you can get ARP to work for egress IP now that OVN isnt handling it anymore?

yep, adding the IP to the bridge generates OVS flows internally. Id rather manage the assignment/removal of an IP address than manually add OVS flows.

For IPv6, it is indeed depreciated if you look at here:
```go
func getEIPBridgeNetlinkAddress(ip net.IP, ifindex int) netlink.Addr {
	return netlink.Addr{
		IPNet:     &net.IPNet{IP: ip, Mask: util.GetIPFullMask(ip)},
		Flags:     getEIPNetlinkAddressFlag(ip),
		Scope:     int(netlink.SCOPE_LINK),
		ValidLft:  getEIPNetlinkAddressValidLft(ip),
		LinkIndex: ifindex,
	}
}

func getEIPNetlinkAddressFlag(ip net.IP) int {
	// isV6?
	if ip.To4() == nil && ip.To16() != nil {
		return unix.IFA_F_NODAD
	}
	return 0
}

func getEIPNetlinkAddressValidLft(ip net.IP) int {
	// isV6?
	if ip.To4() == nil && ip.To16() != nil {
		return math.MaxUint32
	}
	return 0
}

```

martinkennelly commented on test/e2e/egressip.go:
```diff
@@ -1091,6 +1122,9 @@ spec:
 	*/
 	ginkgo.It("Should validate egress IP logic when one pod is managed by more than one egressIP object", func() {
+		if netConfigParams.networkName != types.DefaultNetworkName {
+			ginkgo.Skip("Unsupported for UDNs")
```
because the e2e is heavily tied to CDN internal implementation details.

martinkennelly commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1419,44 +1963,105 @@ func (e *EgressIPController) addEgressNode(node *v1.Node) error {
+	}
+	processedNetworks := make(map[string]struct{})
+	processedNetworks[types.DefaultNetworkName] = struct{}{}
+	for _, namespace := range namespaces {
```
Added as commit on top.

martinkennelly commented on go-controller/pkg/node/gateway_egressip.go:
```diff
@@ -0,0 +1,521 @@
+	} else if ip.To16() != nil {
+		m = e.v6
+	}
+	for _, existingIP := range m {
```
added as commit on top

martinkennelly commented:
Rebase only.
Currenly investigating 2 downstream premerge testing bugs.

martinkennelly commented:
Needed to modify func args `clusterSubnets` from `[]*net.IPNet` to `[]config.CIDRNetworkEntry` :
```go
func CreateDefaultRouteToExternal(nbClient libovsdbclient.Client, clusterRouter, gwRouterName string, clusterSubnets []config.CIDRNetworkEntry) error {
```
in commit "EIP OVN: make controller multi net aware & consider pkt mark"  because of rickys change "Fix predicate for cluster subnet route to gateway router".

martinkennelly commented:
Needed to rebase following merge of https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4868

martinkennelly commented:
Second rebase today :(

Expecting EIP unit tests added by new commit  https://github.com/ovn-kubernetes/ovn-kubernetes/commit/dfa9d8783069285181243c7328b900fd52cffea1 to fail. Ill fix them up tomorrow.

martinkennelly commented:
Starting scale testing. Will report back results.

trozet commented on go-controller/pkg/util/node_annotations.go:
```diff
@@ -104,6 +104,9 @@ const (
 	OVNNodeSecondaryHostEgressIPs = "k8s.ovn.org/secondary-host-egress-ips"
 
+	// OVNNodeBridgeEgressIPs contains the EIP addresses that are assigned to default external bridge linux interface of type OVS.
+	OVNNodeBridgeEgressIPs = "k8s.ovn.org/bridge-egress-ips"
```
I guess we dont.

trozet commented on test/e2e/egressip.go:
```diff
@@ -2712,6 +2713,121 @@ spec:
+	// two pods attached to different namespaces but the same role primary user defined network
+	// One pod is deleted and ensure connectivity for the other pod is ok
+	// The previous pod namespace is deleted and again, ensure connectivity for the other pod is ok
+	ginkgo.It("[OVN network] multiple namespaces sharing a primary networks", func() {
```
nice

martinkennelly commented:
Unit tests  failed because:
```
[38;5;9m• [FAILED] [0.022 seconds][0m
2024-12-11T11:16:03.4169234Z [0mSecondaryNodeNetworkController: UserDefinedPrimaryNetwork Gateway functionality [38;5;9m[1m[It] ensure UDNGateway and VRFManager and IPRulesManager are invoked for Primary UDNs when feature gate is ON[0m
2024-12-11T11:16:03.4171211Z [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/testing/testing.go:18[0m
2024-12-11T11:16:03.4171913Z 
2024-12-11T11:16:03.4173644Z   [38;5;9m[FAILED] Unexpected error:
2024-12-11T11:16:03.4174252Z       <*fmt.wrapError | 0xc000560ac0>: 
2024-12-11T11:16:03.4175470Z       failed to add network to node gateway for network bluenet at node worker1: failed to add the service masquerade chain: /dev/stdin:1:16-29: Error: No such file or directory
2024-12-11T11:16:03.4176594Z       add chain inet ovn-kubernetes udn-mark-0x1003 { comment "bluenet: UDN packet marking" ; }
2024-12-11T11:16:03.4181384Z                      ^^^^^^^^^^^^^^
2024-12-11T11:16:03.4181830Z       
2024-12-11T11:16:03.4182155Z       {
2024-12-11T11:16:03.4185218Z           msg: "failed to add network to node gateway for network bluenet at node worker1: failed to add the service masquerade chain: /dev/stdin:1:16-29: Error: No such file or directory\nadd chain inet ovn-kubernetes udn-mark-0x1003 { comment \"bluenet: UDN packet marking\" ; }\n               ^^^^^^^^^^^^^^\n",
2024-12-11T11:16:03.4187387Z           err: <*fmt.wrapError | 0xc000560aa0>{
2024-12-11T11:16:03.4190078Z               msg: "failed to add the service masquerade chain: /dev/stdin:1:16-29: Error: No such file or directory\nadd chain inet ovn-kubernetes udn-mark-0x1003 { comment \"bluenet: UDN packet marking\" ; }\n               ^^^^^^^^^^^^^^\n",
2024-12-11T11:16:03.4191135Z               err: <*knftables.nftablesError | 0xc000953020>{
2024-12-11T11:16:03.4191672Z                   wrapped: <*exec.ExitError | 0xc000560a60>{
2024-12-11T11:16:03.4192091Z                       ProcessState: {
2024-12-11T11:16:03.4192716Z                           pid: 35489,
2024-12-11T11:16:03.4193114Z                           status: 256,
2024-12-11T11:16:03.4193483Z                           rusage: {
2024-12-11T11:16:03.4194010Z                               Utime: {Sec: 0, Usec: 2031},
2024-12-11T11:16:03.4194542Z                               Stime: {Sec: 0, Usec: 0},
2024-12-11T11:16:03.4195010Z                               Maxrss: 129884,
2024-12-11T11:16:03.4195428Z                               Ixrss: 0,
2024-12-11T11:16:03.4195829Z                               Idrss: 0,
2024-12-11T11:16:03.4196226Z                               Isrss: 0,
2024-12-11T11:16:03.4196636Z                               Minflt: 263,
2024-12-11T11:16:03.4197679Z                               Majflt: 1,
2024-12-11T11:16:03.4198091Z                               Nswap: 0,
2024-12-11T11:16:03.4198513Z                               Inblock: 0,
2024-12-11T11:16:03.4198936Z                               Oublock: 0,
2024-12-11T11:16:03.4199340Z                               Msgsnd: 0,
2024-12-11T11:16:03.4199751Z                               Msgrcv: 0,
2024-12-11T11:16:03.4200166Z                               Nsignals: 0,
2024-12-11T11:16:03.4200565Z                               Nvcsw: 1,
2024-12-11T11:16:03.4200956Z                               Nivcsw: 1,
2024-12-11T11:16:03.4201305Z                           },
2024-12-11T11:16:03.4201750Z                       },
2024-12-11T11:16:03.4206511Z                       Stderr: [47, 100, 101, 118, 47, 115, 116, 100, 105, 110, 58, 49, 58, 49, 54, 45, 50, 57, 58, 32, 69, 114, 114, 111, 114, 58, 32, 78, 111, 32, 115, 117, 99, 104, 32, 102, 105, 108, 101, 32, 111, 114, 32, 100, 105, 114, 101, 99, 116, 111, 114, 121, 10, 97, 100, 100, 32, 99, 104, 97, 105, 110, 32, 105, 110, 101, 116, 32, 111, 118, 110, 45, 107, 117, 98, 101, 114, 110, 101, 116, 101, 115, 32, 117, 100, 110, 45, 109, 97, 114, 107, 45, 48, 120, 49, 48, 48, 51, 32, 123, 32, 99, 111, 109, 109, 101, 110, 116, 32, 34, 98, 108, 117, 101, 110, 101, 116, 58, 32, 85, 68, 78, 32, 112, 97, 99, 107, 101, 116, 32, 109, 97, 114, 107, 105, 110, 103, 34, 32, 59, 32, 125, 10, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 94, 94, 94, 94, 94, 94, 94, 94, 94, 94, 94, 94, 94, 94, 10],
2024-12-11T11:16:03.4207809Z                   },
2024-12-11T11:16:03.4209222Z                   msg: "/dev/stdin:1:16-29: Error: No such file or directory\nadd chain inet ovn-kubernetes udn-mark-0x1003 { comment \"bluenet: UDN packet marking\" ; }\n               ^^^^^^^^^^^^^^\n",
2024-12-11T11:16:03.4209966Z                   errno: 0x2,
2024-12-11T11:16:03.4210219Z               },
2024-12-11T11:16:03.4210420Z           },
2024-12-11T11:16:03.4210592Z       }
2024-12-11T11:16:03.4210847Z   occurred[0m
2024-12-11T11:16:03.4211799Z   [38;5;9mIn [1m[It][0m[38;5;9m at: [1m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/secondary_node_network_controller_test.go:335[0m [38;5;243m@ 12/11/24 11:16:03.414[0m
2024-12-11T11:16:03.4212564Z [38;5;243m------------------------------[0m
```

martinkennelly commented:
### E2E test failure investigation
- [e2e (network-segmentation, noHA, local, dualstack, noSnatGW, 1br, ic-single-node-zones)](https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/12275067320/job/34259952011?pr=4530#logs)
No indication this is caused by this PR currently but suspicious: 
```
2024-12-11T15:09:00.2752640Z   I1211 15:09:00.274859 61210 exec_util.go:59] ExecWithOptions {Command:[/bin/sh -c /bin/sh -c 'echo hostname | nc -u -w 1 172.18.0.3 30434 '] Namespace:udn-services-7032 PodName:udn-client ContainerName:agnhost-container Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:false Quiet:false}
2024-12-11T15:09:00.2754739Z   I1211 15:09:00.274902 61210 util.go:499] >>> kubeConfig: /home/runner/ovn.conf
2024-12-11T15:09:00.2758005Z   I1211 15:09:00.275587 61210 exec_util.go:66] ExecWithOptions: Clientset creation
2024-12-11T15:09:00.2760067Z   I1211 15:09:00.275650 61210 exec_util.go:83] ExecWithOptions: execute(POST https://127.0.0.1:45435/api/v1/namespaces/udn-services-7032/pods/udn-client/exec?command=%2Fbin%2Fsh&command=-c&command=%2Fbin%2Fsh+-c+%27echo+hostname+%7C+nc+-u+-w+1+172.18.0.3+30434+%27&container=agnhost-container&container=agnhost-container&stderr=true&stdout=true)
2024-12-11T15:09:01.3259276Z stdout=
2024-12-11T15:09:01.3259954Z stderr=
2024-12-11T15:09:01.3260506Z err=<nil>
2024-12-11T15:09:01.3262711Z   I1211 15:09:01.325720 61210 network_segmentation_services.go:422] Unexpected error: Failed to verify that Client udn-services-7032/udn-client should connect to NodePort service udn-services-7032/udn-service on 172.18.0.3:30434 (node ovn-worker, other node): 
2024-12-11T15:09:01.3264575Z       <wait.errInterrupted>: 
2024-12-11T15:09:01.3265155Z       timed out waiting for the condition
2024-12-11T15:09:01.3265559Z       {
2024-12-11T15:09:01.3266067Z           cause: <*errors.errorString | 0xc0002792e0>{
2024-12-11T15:09:01.3266537Z               s: "timed out waiting for the condition",
2024-12-11T15:09:01.3266925Z           },
2024-12-11T15:09:01.3267104Z       }
2024-12-11T15:09:01.3268206Z   [38;5;9m[FAILED][0m in [It] - /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/network_segmentation_services.go:422 [38;5;243m@ 12/11/24 15:09:01.326[0m
```

Note how nothing returned in stdout,stderr or err:
```
2024-12-11T15:09:01.3259276Z stdout=
2024-12-11T15:09:01.3259954Z stderr=
2024-12-11T15:09:01.3260506Z err=<nil>
```

martinkennelly commented:
/hold

martinkennelly commented:
/unhold

CDN is ok in LGW mode. Ill follow up with another PR for UDN for LGW and add a test lane for Net seg.

trozet reviewed (approved):
I root caused the lgw UDN egress IP issue, i'll post a fix for it, has nothing to do with this PR.

tssurya commented:
@martinkennelly : thanks for your awesome work here!!
